File: mycollab-scheduler/src/main/java/com/esofthead/mycollab/schedule/email/project/impl/ProjectTaskGroupRelayEmailNotificationActionImpl.java
/**
 * This file is part of mycollab-scheduler.
 *
 * mycollab-scheduler is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-scheduler is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-scheduler.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.schedule.email.project.impl;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Service;

import com.esofthead.mycollab.common.MonitorTypeConstants;
import com.esofthead.mycollab.common.domain.SimpleRelayEmailNotification;
import com.esofthead.mycollab.common.i18n.GenericI18Enum;
import com.esofthead.mycollab.configuration.SiteConfiguration;
import com.esofthead.mycollab.core.utils.StringUtils;
import com.esofthead.mycollab.module.mail.MailUtils;
import com.esofthead.mycollab.module.project.ProjectLinkGenerator;
import com.esofthead.mycollab.module.project.ProjectResources;
import com.esofthead.mycollab.module.project.ProjectTypeConstants;
import com.esofthead.mycollab.module.project.domain.SimpleMilestone;
import com.esofthead.mycollab.module.project.domain.SimpleProject;
import com.esofthead.mycollab.module.project.domain.SimpleProjectMember;
import com.esofthead.mycollab.module.project.domain.SimpleTaskList;
import com.esofthead.mycollab.module.project.i18n.TaskGroupI18nEnum;
import com.esofthead.mycollab.module.project.service.MilestoneService;
import com.esofthead.mycollab.module.project.service.ProjectMemberService;
import com.esofthead.mycollab.module.project.service.ProjectService;
import com.esofthead.mycollab.module.project.service.ProjectTaskListService;
import com.esofthead.mycollab.module.user.AccountLinkGenerator;
import com.esofthead.mycollab.module.user.domain.SimpleUser;
import com.esofthead.mycollab.module.user.service.UserService;
import com.esofthead.mycollab.schedule.email.ItemFieldMapper;
import com.esofthead.mycollab.schedule.email.MailContext;
import com.esofthead.mycollab.schedule.email.format.FieldFormat;
import com.esofthead.mycollab.schedule.email.format.html.TagBuilder;
import com.esofthead.mycollab.schedule.email.project.ProjectTaskGroupRelayEmailNotificationAction;
import com.esofthead.mycollab.spring.ApplicationContextUtil;
import com.hp.gagawa.java.elements.A;
import com.hp.gagawa.java.elements.Img;
import com.hp.gagawa.java.elements.Span;

/**
 * 
 * @author MyCollab Ltd.
 * @since 1.0
 * 
 */
@Service
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
public class ProjectTaskGroupRelayEmailNotificationActionImpl extends
		SendMailToAllMembersAction<SimpleTaskList> implements
		ProjectTaskGroupRelayEmailNotificationAction {

	private static Logger log = LoggerFactory
			.getLogger(ProjectTaskGroupRelayEmailNotificationActionImpl.class);

	@Autowired
	private ProjectTaskListService projectTaskListService;

	@Autowired
	private ProjectService projectService;

	@Autowired
	private ProjectMemberService projectMemberService;

	private static final ProjectFieldNameMapper mapper = new ProjectFieldNameMapper();

	@Override
	protected String getItemName() {
		return StringUtils.trim(bean.getName(), 100);
	}

	@Override
	protected String getCreateSubject(MailContext<SimpleTaskList> context) {
		return context.getMessage(TaskGroupI18nEnum.MAIL_CREATE_ITEM_SUBJECT,
				bean.getProjectName(), context.getChangeByUserFullName(),
				getItemName());
	}

	@Override
	protected String getUpdateSubject(MailContext<SimpleTaskList> context) {
		return context.getMessage(TaskGroupI18nEnum.MAIL_UPDATE_ITEM_SUBJECT,
				bean.getProjectName(), context.getChangeByUserFullName(),
				getItemName());
	}

	@Override
	protected String getCommentSubject(MailContext<SimpleTaskList> context) {
		return context.getMessage(TaskGroupI18nEnum.MAIL_COMMENT_ITEM_SUBJECT,
				bean.getProjectName(), context.getChangeByUserFullName(),
				getItemName());
	}

	@Override
	protected ItemFieldMapper getItemFieldMapper() {
		return mapper;
	}

	@Override
	protected SimpleTaskList getBeanInContext(
			MailContext<SimpleTaskList> context) {
		return projectTaskListService.findById(context.getTypeid(),
				context.getSaccountid());
	}

	@Override
	protected void buildExtraTemplateVariables(
			MailContext<SimpleTaskList> context) {
		List<Map<String, String>> listOfTitles = new ArrayList<Map<String, String>>();

		SimpleRelayEmailNotification emailNotification = context
				.getEmailNotification();

		SimpleProject relatedProject = projectService.findById(
				bean.getProjectid(), emailNotification.getSaccountid());

		HashMap<String, String> currentProject = new HashMap<String, String>();
		currentProject.put("displayName", relatedProject.getName());
		currentProject.put(
				"webLink",
				ProjectLinkGenerator.generateProjectFullLink(siteUrl,
						bean.getProjectid()));

		listOfTitles.add(currentProject);

		String summary = bean.getName();
		String summaryLink = ProjectLinkGenerator
				.generateTaskGroupPreviewFullLink(siteUrl, bean.getProjectid(),
						bean.getId());

		String avatarId = "";

		SimpleProjectMember projectMember = projectMemberService
				.findMemberByUsername(emailNotification.getChangeby(),
						bean.getProjectid(), emailNotification.getSaccountid());
		if (projectMember != null) {
			avatarId = projectMember.getMemberAvatarId();
		}
		Img userAvatar = new Img("", SiteConfiguration.getAvatarLink(avatarId,
				16));
		userAvatar.setWidth("16");
		userAvatar.setHeight("16");
		userAvatar.setStyle("display: inline-block; vertical-align: top;");

		String makeChangeUser = userAvatar.toString()
				+ emailNotification.getChangeByUserFullName();

		if (MonitorTypeConstants.CREATE_ACTION.equals(emailNotification
				.getAction())) {
			contentGenerator
					.putVariable("actionHeading", context.getMessage(
							TaskGroupI18nEnum.MAIL_CREATE_ITEM_HEADING,
							makeChangeUser));
		} else if (MonitorTypeConstants.UPDATE_ACTION.equals(emailNotification
				.getAction())) {
			contentGenerator
					.putVariable("actionHeading", context.getMessage(
							TaskGroupI18nEnum.MAIL_UPDATE_ITEM_HEADING,
							makeChangeUser));
		} else if (MonitorTypeConstants.ADD_COMMENT_ACTION
				.equals(emailNotification.getAction())) {
			contentGenerator.putVariable("actionHeading", context
					.getMessage(TaskGroupI18nEnum.MAIL_COMMENT_ITEM_HEADING,
							makeChangeUser));
		}
		contentGenerator.putVariable("titles", listOfTitles);
		contentGenerator.putVariable("summary", summary);
		contentGenerator.putVariable("summaryLink", summaryLink);
	}

	public static class ProjectFieldNameMapper extends ItemFieldMapper {

		public ProjectFieldNameMapper() {

			put("name", TaskGroupI18nEnum.FORM_NAME_FIELD, true);

			put("owner", new AssigneeFieldFormat("owner",
					GenericI18Enum.FORM_ASSIGNEE));
			put("status", TaskGroupI18nEnum.FORM_STATUS);

			put("milestoneid", new MilestoneFieldFormat("milestoneid",
					TaskGroupI18nEnum.FORM_PHASE_FIELD, true));

			put("description", GenericI18Enum.FORM_DESCRIPTION, true);

		}
	}

	public static class AssigneeFieldFormat extends FieldFormat {

		public AssigneeFieldFormat(String fieldName, Enum<?> displayName) {
			super(fieldName, displayName);
		}

		@Override
		public String formatField(MailContext<?> context) {
			SimpleTaskList tasklist = (SimpleTaskList) context.getWrappedBean();
			if (tasklist.getOwner() != null) {
				String userAvatarLink = MailUtils.getAvatarLink(
						tasklist.getOwnerAvatarId(), 16);
				Img img = TagBuilder.newImg("avatar", userAvatarLink);

				String userLink = AccountLinkGenerator
						.generatePreviewFullUserLink(
								MailUtils.getSiteUrl(tasklist.getSaccountid()),
								tasklist.getOwner());
				A link = TagBuilder.newA(userLink, tasklist.getOwnerFullName());
				return TagBuilder.newLink(img, link).write();
			} else {
				return new Span().write();
			}

		}

		@Override
		public String formatField(MailContext<?> context, String value) {
			if (value == null || "".equals(value)) {
				return new Span().write();
			}

			UserService userService = ApplicationContextUtil
					.getSpringBean(UserService.class);
			SimpleUser user = userService.findUserByUserNameInAccount(value,
					context.getUser().getAccountId());
			if (user != null) {
				String userAvatarLink = MailUtils.getAvatarLink(
						user.getAvatarid(), 16);
				String userLink = AccountLinkGenerator
						.generatePreviewFullUserLink(
								MailUtils.getSiteUrl(user.getAccountId()),
								user.getUsername());
				Img img = TagBuilder.newImg("avatar", userAvatarLink);
				A link = TagBuilder.newA(userLink, user.getDisplayName());
				return TagBuilder.newLink(img, link).write();
			}
			return value;
		}
	}

	public static class MilestoneFieldFormat extends FieldFormat {

		public MilestoneFieldFormat(String fieldName, Enum<?> displayName,
				boolean isColSpan) {
			super(fieldName, displayName, isColSpan);
		}

		@Override
		public String formatField(MailContext<?> context) {
			SimpleTaskList tasklist = (SimpleTaskList) context.getWrappedBean();
			if (tasklist.getMilestoneid() != null) {
				String milestoneIconLink = ProjectResources
						.getResourceLink(ProjectTypeConstants.MILESTONE);
				Img img = TagBuilder.newImg("icon", milestoneIconLink);

				String milestoneLink = ProjectLinkGenerator
						.generateMilestonePreviewFullLink(context.getSiteUrl(),
								tasklist.getProjectid(),
								tasklist.getMilestoneid());
				A link = TagBuilder.newA(milestoneLink,
						tasklist.getMilestoneName());
				return TagBuilder.newLink(img, link).write();
			} else {
				return new Span().write();
			}

		}

		@Override
		public String formatField(MailContext<?> context, String value) {
			if (value == null || "".equals(value)) {
				return new Span().write();
			}

			try {
				int milestoneId = Integer.parseInt(value);
				MilestoneService milestoneService = ApplicationContextUtil
						.getSpringBean(MilestoneService.class);
				SimpleMilestone milestone = milestoneService.findById(
						milestoneId, context.getUser().getAccountId());

				if (milestone != null) {
					String milestoneIconLink = ProjectResources
							.getResourceLink(ProjectTypeConstants.MILESTONE);
					Img img = TagBuilder.newImg("icon", milestoneIconLink);

					String milestoneLink = ProjectLinkGenerator
							.generateMilestonePreviewFullLink(
									context.getSiteUrl(),
									milestone.getProjectid(), milestone.getId());
					A link = TagBuilder
							.newA(milestoneLink, milestone.getName());
					return TagBuilder.newLink(img, link).write();
				}
			} catch (Exception e) {
				log.error("Error", e);
			}

			return value;
		}

	}
}

File: mycollab-scheduler/src/main/scala/com/esofthead/mycollab/schedule/email/project/impl/ProjectTaskGroupRelayEmailNotificationActionImpl.scala
/**
 * This file is part of mycollab-scheduler.
 *
 * mycollab-scheduler is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-scheduler is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-scheduler.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.schedule.email.project.impl

import com.esofthead.mycollab.common.MonitorTypeConstants
import com.esofthead.mycollab.common.domain.SimpleRelayEmailNotification
import com.esofthead.mycollab.common.i18n.GenericI18Enum
import com.esofthead.mycollab.core.utils.StringUtils
import com.esofthead.mycollab.html.{FormatUtils, LinkUtils}
import com.esofthead.mycollab.module.mail.MailUtils
import com.esofthead.mycollab.module.project.domain._
import com.esofthead.mycollab.module.project.i18n.TaskGroupI18nEnum
import com.esofthead.mycollab.module.project.service.{MilestoneService, ProjectService, ProjectTaskListService}
import com.esofthead.mycollab.module.project.{ProjectLinkGenerator, ProjectResources, ProjectTypeConstants}
import com.esofthead.mycollab.module.user.AccountLinkGenerator
import com.esofthead.mycollab.module.user.domain.SimpleUser
import com.esofthead.mycollab.module.user.service.UserService
import com.esofthead.mycollab.schedule.email.format.{FieldFormat, WebItem}
import com.esofthead.mycollab.schedule.email.project.ProjectTaskGroupRelayEmailNotificationAction
import com.esofthead.mycollab.schedule.email.{ItemFieldMapper, MailContext}
import com.esofthead.mycollab.spring.ApplicationContextUtil
import com.hp.gagawa.java.elements.{A, Img, Span, Text}
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Service

/**
 * @author MyCollab Ltd.
 * @since 4.6.0
 */
@Service
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
class ProjectTaskGroupRelayEmailNotificationActionImpl extends SendMailToAllMembersAction[SimpleTaskList] with
ProjectTaskGroupRelayEmailNotificationAction {

    private val LOG = LoggerFactory.getLogger(classOf[ProjectTaskGroupRelayEmailNotificationActionImpl])

    @Autowired var projectTaskListService: ProjectTaskListService = _

    @Autowired var projectService: ProjectService = _

    private val mapper = new ProjectFieldNameMapper

    protected def getItemName: String = StringUtils.trim(bean.getName, 100)

    protected def getCreateSubject(context: MailContext[SimpleTaskList]): String = context.getMessage(
        TaskGroupI18nEnum.MAIL_CREATE_ITEM_SUBJECT, bean.getProjectName, context.getChangeByUserFullName, getItemName)

    protected def getUpdateSubject(context: MailContext[SimpleTaskList]): String = context.getMessage(
        TaskGroupI18nEnum.MAIL_UPDATE_ITEM_SUBJECT, bean.getProjectName, context.getChangeByUserFullName, getItemName)

    protected def getCommentSubject(context: MailContext[SimpleTaskList]): String = context.getMessage(
        TaskGroupI18nEnum.MAIL_COMMENT_ITEM_SUBJECT, bean.getProjectName, context.getChangeByUserFullName, getItemName)

    protected def getItemFieldMapper: ItemFieldMapper = mapper

    protected def getBeanInContext(context: MailContext[SimpleTaskList]): SimpleTaskList = projectTaskListService.findById(
        context.getTypeid.toInt, context.getSaccountid)

    protected def buildExtraTemplateVariables(context: MailContext[SimpleTaskList]) {
        val emailNotification: SimpleRelayEmailNotification = context.getEmailNotification
        val relatedProject: SimpleProject = projectService.findById(bean.getProjectid, emailNotification.getSaccountid)

        val currentProject = new WebItem(relatedProject.getName, ProjectLinkGenerator.generateProjectFullLink(siteUrl,
            bean.getProjectid))

        val summary: String = bean.getName
        val summaryLink: String = ProjectLinkGenerator.generateTaskGroupPreviewFullLink(siteUrl, bean.getProjectid, bean.getId)
        val projectMember: SimpleProjectMember = projectMemberService.findMemberByUsername(emailNotification.getChangeby,
            bean.getProjectid, emailNotification.getSaccountid)

        val avatarId: String = if (projectMember != null) projectMember.getMemberAvatarId else ""
        val userAvatar: Img = LinkUtils.newAvatar(avatarId)

        val makeChangeUser: String = userAvatar.toString + emailNotification.getChangeByUserFullName
        val headerEnum: Enum[_] = emailNotification.getAction match {
            case MonitorTypeConstants.CREATE_ACTION => TaskGroupI18nEnum.MAIL_CREATE_ITEM_HEADING
            case MonitorTypeConstants.UPDATE_ACTION => TaskGroupI18nEnum.MAIL_UPDATE_ITEM_HEADING
            case MonitorTypeConstants.ADD_COMMENT_ACTION => TaskGroupI18nEnum.MAIL_COMMENT_ITEM_HEADING
        }

        contentGenerator.putVariable("actionHeading", context.getMessage(headerEnum, makeChangeUser))
        contentGenerator.putVariable("titles", List(currentProject))
        contentGenerator.putVariable("summary", summary)
        contentGenerator.putVariable("summaryLink", summaryLink)
    }

    class ProjectFieldNameMapper extends ItemFieldMapper {
        put(TaskList.Field.name, TaskGroupI18nEnum.FORM_NAME_FIELD, isColSpan = true)
        put(TaskList.Field.owner, new AssigneeFieldFormat(TaskList.Field.owner.name, GenericI18Enum.FORM_ASSIGNEE))
        put(TaskList.Field.status, TaskGroupI18nEnum.FORM_STATUS)
        put(TaskList.Field.milestoneid, new MilestoneFieldFormat(TaskList.Field.milestoneid.name, TaskGroupI18nEnum.FORM_PHASE_FIELD, true))
        put(TaskList.Field.description, GenericI18Enum.FORM_DESCRIPTION, isColSpan = true)
    }

    class AssigneeFieldFormat(fieldName: String, displayName: Enum[_]) extends FieldFormat(fieldName, displayName) {
        def formatField(context: MailContext[_]): String = {
            val taskList: SimpleTaskList = context.getWrappedBean.asInstanceOf[SimpleTaskList]
            if (taskList.getOwner != null) {
                val userAvatarLink: String = MailUtils.getAvatarLink(taskList.getOwnerAvatarId, 16)
                val img: Img = FormatUtils.newImg("avatar", userAvatarLink)
                val userLink: String = AccountLinkGenerator.generatePreviewFullUserLink(MailUtils.getSiteUrl(taskList.getSaccountid),
                    taskList.getOwner)
                val link: A = FormatUtils.newA(userLink, taskList.getOwnerFullName)
                FormatUtils.newLink(img, link).write
            }
            else {
                new Span().write
            }
        }

        def formatField(context: MailContext[_], value: String): String = {
            if (org.apache.commons.lang3.StringUtils.isBlank(value)) {
                new Span().write
            } else {
                val userService: UserService = ApplicationContextUtil.getSpringBean(classOf[UserService])
                val user: SimpleUser = userService.findUserByUserNameInAccount(value, context.getUser.getAccountId)
                if (user != null) {
                    val userAvatarLink: String = MailUtils.getAvatarLink(user.getAvatarid, 16)
                    val userLink: String = AccountLinkGenerator.generatePreviewFullUserLink(MailUtils.getSiteUrl(user.getAccountId),
                        user.getUsername)
                    val img: Img = FormatUtils.newImg("avatar", userAvatarLink)
                    val link: A = FormatUtils.newA(userLink, user.getDisplayName)
                    FormatUtils.newLink(img, link).write
                } else
                    value
            }
        }
    }

    class MilestoneFieldFormat(fieldName: String, displayName: Enum[_], isColSpan: Boolean) extends FieldFormat(fieldName,
        displayName, isColSpan) {
        def formatField(context: MailContext[_]): String = {
            val taskList: SimpleTaskList = context.wrappedBean.asInstanceOf[SimpleTaskList]
            if (taskList.getMilestoneid != null) {
                val img: Text = new Text(ProjectResources.getFontIconHtml(ProjectTypeConstants.MILESTONE));
                val milestoneLink: String = ProjectLinkGenerator.generateMilestonePreviewFullLink(context.siteUrl, taskList
                    .getProjectid, taskList.getMilestoneid)
                val link: A = FormatUtils.newA(milestoneLink, taskList.getMilestoneName)
                FormatUtils.newLink(img, link).write
            }
            else {
                new Span().write
            }
        }

        def formatField(context: MailContext[_], value: String): String = {
            if (org.apache.commons.lang3.StringUtils.isBlank(value)) {
                return new Span().write
            }
            try {
                val milestoneId: Int = value.toInt
                val milestoneService: MilestoneService = ApplicationContextUtil.getSpringBean(classOf[MilestoneService])
                val milestone: SimpleMilestone = milestoneService.findById(milestoneId, context.getUser.getAccountId)
                if (milestone != null) {
                    val img: Text = new Text(ProjectResources.getFontIconHtml(ProjectTypeConstants.MILESTONE))
                    val milestoneLink: String = ProjectLinkGenerator.generateMilestonePreviewFullLink(context.siteUrl,
                        milestone.getProjectid, milestone.getId)
                    val link: A = FormatUtils.newA(milestoneLink, milestone.getName)
                    return FormatUtils.newLink(img, link).write
                }
            }
            catch {
                case e: Exception => LOG.error("Error", e)
            }
            value
        }
    }

}
